magma/value_utils.py

```diff
@@ -0,0 +1,107 @@
+import dataclasses
+import typing
+
+from magma.ref import Ref, TupleRef, ArrayRef
+from magma.type_utils import isprotocol, isdigital, isbits, isarray, istuple
+
+
+def _unwrap_protocol_value(value):
+    T = type(value)
+    if isprotocol(T):
+        return value._get_magma_value_()
+    return value
+
+
+@dataclasses.dataclass(frozen=True)
+class _ValueWrapper:
+    raw: typing.Any
+    children: typing.List[typing.Any]
+    name: str
+
+
+def _wrap_value(value):
+    if isinstance(value, _ValueWrapper):
+        return value
+    value = _unwrap_protocol_value(value)
+    T = type(value)
+
+    if isdigital(T):
+        return _ValueWrapper(value, [], "Digital")
+    if isbits(T):
+        return _ValueWrapper(value, [], "Bits")
+    if isarray(T):
+        return _ValueWrapper(value, [t for t in value], "Array")
+    if istuple(T):
+        return _ValueWrapper(value, [t for t in value], "Tuple")
+
+    raise NotImplementedError(T)
+
+
+class ValueVisitor:
+    def visit(self, value):
+        wrapped = _wrap_value(value)
+        method = 'visit_' + wrapped.name
+        visitor = getattr(self, method, self.generic_visit)
+        return visitor(value)
+
+    def generic_visit(self, value):
+        wrapped = _wrap_value(value)
+        for child in wrapped.children:
+            self.visit(child)
+
+
+@dataclasses.dataclass(frozen=True)
+class Selector:
+    child: 'Selector'
+
+    def select(self, value):
+        value = self._select(value)
+        if self.child is None:
+            return value
+        return self.child.select(value)
+
+    def _select(self, value):
+        return value
+
+    def _child_str(self):
+        return "" if self.child is None else str(self.child)
```

Comment:
Should this define __str__ method that prepends the value name before the child names?

Reply:
Actually, never mind it looks like this is only used in the case when there's no children (value is None)